bugsnag/bugsnag-react-native
# Fix possible release stage mismatch between sessions and error reports

## Goal

It is possible for the native release stage to not match the eventual actual release stage set on the JS side, and sessions could have a different release stage from subsequent errors. This change disables the native logic for automatically reporting the first session replacing it with a call to `startSession` from `startWithOptions`.

This change assumes that the React Native layer is the primary way to interact with the application, and when this is not the case, a different session tracking setup should be used instead of the automatic session detection provided (which should be added to the documentation for session tracking). 

This is also required to update to bugsnag-cocoa@5.16.2 as discussed in https://github.com/bugsnag/bugsnag-cocoa/pull/295#discussion_r202384805.

## Changeset

### Changed

* Updated bugsnag-cocoa to v5.16.2
* Disabled automatic session tracking in cocoa's `startWithConfiguration` and android's `init` methods
* Added an automatically tracked session in `startWithOptions` if automatic
  session tracking is enabled in JS

## Tests

Manually tested:

* Starting an application with the default release stage settings to confirm a single session is sent on iOS and Android
* Starting an application with a custom release stage set in JS to confirm a single session is sent on iOS and Android
* Causing handled and unhandled errors to ensure session counts and release stages match between sessions and error reports on iOS and Android
* Disabling automatic session tracking to ensure no sessions were sent at launch/reload
* Disabling automatic session tracking and manually sending a session to ensure a single session is sent
* Manually sending sessions with automatic session tracking enabled to ensure that both automatic and manual sessions are recorded

## Discussion

### Alternative Approaches

This implementation sends a session each time Bugsnag is initialized on the JS side, regardless of whether the native side has been running previously. My initial implementation would check whether the native side already has session tracking enabled and only send a session if it was was not natively enabled, however, this led to an inconsistent experience when soft reloading the JS components but not subsequently getting a new session sent. 

### Outstanding Questions

While this removes the possibility of sending a session for the wrong release stage, it raises the possibility of reporting an error with no session set at all. Any native error which occurs before JS-side initialization occurs will not have an assigned session. Is there a better strategy for mitigating this? We could potentially recommend setting release stage during native setup if using customized settings, but that may require information not available before the JS runtime is initialized.

### Linked issues

Related to bugsnag/bugsnag-cocoa#295

## Review



For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language
